### PR TITLE
feat(http): Add pluggable HTTP cache implementation

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Adds new classes, without using them anywhere for now

## Context

- Extracted from the larger PR: #27966

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/http/cache/abstract-http-cache-provider.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    if (resp.statusCode === 200) {
+      const etag = resp.headers?.['etag'];
+      const lastModified = resp.headers?.['last-modified'];
+
+      HttpCacheStats.incRemoteMisses(url);
+
+      const httpResponse = copyResponse(resp, true) as unknown as Record<
+        string,
+        unknown
+      >;
```
Comment (viceice) [inline lib/util/http/cache/abstract-http-cache-provider.ts:55]: Why this ugly cast?

File: lib/util/http/types.ts [hunk 2]
```diff
@@ -89,3 +91,11 @@ export interface HttpResponse<T = string> {
 
 export type Task<T> = () => Promise<T>;
 export type GotTask<T> = Task<HttpResponse<T>>;
+
+// @deprecated
```
Comment (viceice) [inline lib/util/http/types.ts:95]: use tsdoc style for deprecated

**Review by viceice — COMMENTED** [review]
use tsdoc comment style, otherwise LGTM

File: lib/util/http/cache/abstract-http-cache-provider.ts (hunk 1 above)
Comment (zharinov) [reply lib/util/http/cache/abstract-http-cache-provider.ts]: Never mind, just realized I've already made changes but forgot to remove type cast

File: lib/util/http/cache/schema.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { z } from 'zod';
+
+const invalidFieldsMsg =
+  'Cache object should have `etag` or `lastModified` fields';
+
+export const HttpCacheSchema = z
+  .object({
+    timeStamp: z.string().optional(),
```
Comment (rarkins) [inline lib/util/http/cache/schema.ts:8]: Is this temporary and needs a //TODO: remove later?

**Review by viceice — APPROVED** [review]

**Pull request merged**